src/parser/common.rs

```diff
@@ -373,4 +415,65 @@ mod tests {
             .collect()
         )
     }
+
+    #[test]
+    fn test_parse_matrix() {
+        let input = "\t1/sqrt(2), 1/sqrt(2)\n\t1/sqrt(2), -1/sqrt(2)";
+        let lexed = lex(input).unwrap();
+        let (remainder, matrix) = parse_matrix(&lexed).unwrap();
+        assert!(
+            remainder.is_empty(),
+            "expected remainder to be empty, got {:?}",
+            remainder
+        );
+        assert_eq!(matrix.len(), 2);
+    }
+
+    #[test]
+    fn test_parse_permutation() {
+        let input = "\t0, 1, 2, 3, 4, 5, 7, 6";
+        let lexed = lex(input).unwrap();
+        let (remainder, matrix) = parse_permutation(&lexed).unwrap();
+        assert!(
+            remainder.is_empty(),
+            "expected remainder to be empty, got {:?}",
+            remainder
+        );
+        assert_eq!(
+            matrix,
+            vec![vec![
+                Expression::Number(real!(0.0)),
```

Comment:
Ideally, these would parse as non-negative integers. Not sure how feasible that is, though.

Reply:
This is where I struggled the most. I looked into adding integers as a valid expression, but that requires trickles into a lot of other places. Once I got to `evaluate()` I realized it would have to ultimately be converted into a float and that just felt weird.

I'll take any suggestions though.

Reply:
An integer is a token already - don't need to consider it an expression. See `parse_expression` to see that it's handling integer tokens internally, which you can do directly

Reply:
Ah, I missed that. Thank you @kalzoo. That was the one thing I really disliked about the implementation I have.